### ISSUE TITLE
Foreach type check: false positive with `: array` hint.

### ISSUE THREAD

**Issue by cburschka-pwc** [kalessil/phpinspectionsea#640]
Moved from https://youtrack.jetbrains.com/issue/WI-38729:

PHP hints don't support any type more specific than `array`, but you can use more detailed ones such as `Class[]` or `array[]` or `string[][]` in the PHPDoc comment. PhpStorm recognizes these types, such as in the following file:

```php
class NestedArrayTest {
  /**
   * Returns a nested array.
   *
   * @return string[][]
   */
  public function nestedArray() {
    return [['']];
  }

  /**
   * Loops over a nested array.
   */
  public function iterate() {
    foreach ($this->nestedArray() as $row) {
      // This does not trigger a warning.
      foreach ($row as $cell) {
        print strlen($cell);
      }
    }
  }
}
```

However, if you add both a PHP hint and a comment hint, the inspection gets a false positive:

```php
class NestedArrayTest {
  /**
   * Returns a nested array.
   *
   * @return string[][]
   */
  public function nestedArray(): array {
    return [['']];
  }

  /**
   * Loops over a nested array.
   */
  public function iterate() {
    foreach ($this->nestedArray() as $row) {
      // This triggers a type inspection warning:
      foreach ($row as $cell) {
        print strlen($cell);
      }
    }
  }
}
```

This is because the return value is typed as `string[][]|array`, which results in `$row` being typed `string[]|mixed`, and `mixed` is not necessarily iterable.

If possible, I'd suggest that the `: array` hint should only be added to the alternatives if there isn't already a `[]` type in the PHPDoc comment, since the latter is more specific than the former.

**Comment by kalessil** [comment]
Thank you for reporting @cburschka-pwc, fixed!
